[backend] Add Zulip Backend

vchrombie opened the pull request:
Fixes #630

vchrombie commented:
The PR is not completed yet and needs a lot of improvements. I am still figuring out how to use offset based method (I have a doubt if it is applicable :thinking:).

Some of the channels where we can test this are
1. https://python.zulipchat.com/
2. https://anitab-org.zulipchat.com/

The usage of this backend will be
```
perceval zulip 'https://python.zulipchat.com/' 'importlib' -e 'BOT_EMAIL_ADDRESS' -t 'BOT_API_KEY'
```

You can generate the BOT_EMAIL_ADDRESS and BOT_API_KEY from the Zulip chat itself (you need to join the chat). 
Click on the :gear: >> Settings >> Your bots >> Add a new bot


I have made a small script that fetches a fixed number of messages, not in an iterative manner.
[https://gist.github.com/vchrombie/bfd6d967e03bbe28cf4241995fb4b91b](https://gist.github.com/vchrombie/bfd6d967e03bbe28cf4241995fb4b91b#file-script-py)

valeriocos commented:
Hi @vchrombie , sorry for the late reply on this PR! I'm on it

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,131 @@
+
+    CATEGORIES = [CATEGORY_MESSAGE]
+
+    def __init__(self, url, stream, api_token, email, origin, tag=None, archive=None, ssl_verify=True):
```
Origin shouldn't be declared in __init__, since it can be derived from the url and stream. The origin is used to identify a repository.
Line 45 and 48 are duplicated.
The url should be stored a instance attribute, since it will be passed then to the client

Suggested changes below:
```
-    def __init__(self, url, stream, api_token, email, origin, tag=None, archive=None, ssl_verify=True):
-        super().__init__(origin, tag, archive, ssl_verify)
-        origin = url
-
-        super.__init__(origin, tag=tag, archive=archive, ssl_verify=True)
+    def __init__(self, url, stream, api_token, email, tag=None, archive=None, ssl_verify=True):
+        origin = urijoin(url, stream)
+        super().__init__(origin, tag=tag, archive=archive, ssl_verify=ssl_verify)
+        self.url = url
         self.stream = stream
         self.api_token = api_token
         self.email = email
 
         self.client = None
```

vchrombie replied:
Thanks, I was a bit confused about this. I got an idea about it now. :smile: 

Also, I had one more doubt about the `origin`. 
The original URL of the stream is https://python.zulipchat.com/#narrow/stream/187177-importlib.
The `origin` which we generates here is `https://python.zulipchat.com/importlib`.

I was just having second thoughts about this now.

valeriocos replied:
> Thanks, I was a bit confused about this. I got an idea about it now. smile

You're welcome!

I see that `https://python.zulipchat.com/#narrow/stream/importlib` is equivalent to https://python.zulipchat.com/#narrow/stream/187177-importlib. We could set the origin with the first option, which is easier to build, WDYT?

vchrombie replied:
Okay, perfect. :+1:

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,131 @@
+
+        self.client = None
+
+    def fetch(self, category=CATEGORY_MESSAGE):
```
The fetch signature misses the offset. The default offset should be declared as constant, which value should be the first offset possible (e.g., DEFAULT_OFFSET = 1)

```
-    def fetch(self, category=CATEGORY_MESSAGE):
+    def fetch(self, category=CATEGORY_MESSAGE, offset=DEFAULT_OFFSET):
         """Fetch the message from the stream."""
-        kwargs = {}
+
+        kwargs = {"offset": offset}
         items = super().fetch(category, **kwargs)
```

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,131 @@
+        return items
+
+    def fetch_items(self, category, **kwargs):
+        """Fetch the messages."""
```
It could make sense to have a specific method to extract the messages retrieved

```
         """Fetch the messages."""
 
         anchor = 'oldest'
-        raw_message = self.client.get_message(anchor)
-        print(raw_message)
+        raw_message = self.client.get_messages(anchor)
+        messages = self.parse_messages(raw_message)
+        for msg in messages:
+            yield msg
+
+    def parse_messages(self, raw_messages):
+        result = json.loads(raw_messages)
+
+        return result.get('messages', None)
```

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,131 @@
+        anchor = 'oldest'
+        raw_message = self.client.get_message(anchor)
+        print(raw_message)
+
```
The methods to complete the metadata perceval info are missing. Find below possible implementations for them. I guess that Zulip allows to retrieve all history of a channel from the older message to the newest, if this is correct `has_resuming` should return True, otherwise `False` (as similarly done for [Slack](https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/slack.py#L170), which doesn't allow to fetch all history)

```
+    @classmethod
+    def has_archiving(cls):
+        """Returns whether it supports archiving items on the fetch process.
+
+        :returns: this backend supports items archive
+        """
+        return True
+
+    @classmethod
+    def has_resuming(cls):
+        """Returns whether it supports to resume the fetch process.
+
+        :returns: this backend does not support items resuming
+        """
+        return True
+
+    @staticmethod
+    def metadata_id(item):
+        """Extracts the identifier of a Zulip item."""
+
+        return str(item['id'])
+
+    @staticmethod
+    def metadata_updated_on(item):
+        """Extracts and coverts the update time from a Zulip item.
+
+        The timestamp is extracted from 'ts' field and converted
+        to a UNIX timestamp.
+
+        :param item: item generated by the backend
+
+        :returns: a UNIX timestamp
+        """
+        ts = float(item['timestamp'])
+
+        return ts
```

valeriocos replied:
I see that I didn't copy the diff for the method `metadata_category(item):
```
@@ -160,6 +160,15 @@ class Zulip(Backend):
         ts = float(item['timestamp'])
         return ts
 
+    @staticmethod
+    def metadata_category(item):
+        """Extracts the category from a Zulip item.
+
+        This backend only generates one type of item which is
+        'message'.
+        """
+        return CATEGORY_MESSAGE
```

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,131 @@
+    def _init_client(self, from_archive=False):
+        """Init client"""
+
+        return ZulipClient(self.url, self.stream, self.api_token,
```
The param `archive` isn't passed to the client.

Suggested change
```
     def _init_client(self, from_archive=False):
         """Init client"""
 
         return ZulipClient(self.url, self.stream, self.api_token,
-                           self.email, from_archive, self.ssl_verify)
+                           self.email, archive=self.archive, from_archive=from_archive, ssl_verify=self.ssl_verify)
```

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,131 @@
+                 archive=None, from_archive=False, ssl_verify=True):
+        self.stream = stream
+        self.api_token = api_token
+        self.email = email
```
url should be an instance variable, since used later on by the client methods

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,131 @@
+        self.api_token = api_token
+        self.email = email
+
+        super().__init__(url,
```
This could be on a single line

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,131 @@
+        response = super().fetch(url, payload)
+        return response
+
+    def get_message(self, anchor):
```
typo, missing s at the end of the method -> get_messages

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,131 @@
+        return response
+
+    def get_message(self, anchor):
+        request = {
```
To get data from this API we need to pass the token and email. Looking at the [zulip wrapper](https://github.com/zulip/python-zulip-api/blob/16737e4ed3ecd9dc6b1b25f4566f3477ed96b089/zulip/zulip/__init__.py#L419), they rely on HTTPBasicAuth. We can leverage on the param `auth` provided by the generic Perceval fetch method.

Suggested change:
```
+        params = {
             'anchor': '{}'.format(anchor),
             'num_before': '0',
             'num_after': '2',
             'apply_markdown': 'false',
             'narrow': json.dumps([{'operator': 'stream', 'operand': '{}'.format(self.stream)}])
         }
-        path = urijoin(url, "/api/v1/messages")
-        payload = {'params': request}
+        path = urijoin(self.url, "/api/v1/messages")
+        r = self.fetch(path, payload=params, auth=(self.email, self.api_token))
 
-        return self.fetch(path, payload)
+        return r.text
```

vchrombie replied:
Thanks for the suggestion @valeriocos.

I tried this, ended up with an error
```
RuntimeError: fetch() got an unexpected keyword argument 'auth'
```

I think it is because the argument `auth`. I resolved it with this
```
-    def fetch(self, url, payload=None, headers=None):
+    def fetch(self, url, payload=None, headers=None, auth=None):
```
The previous error was not persistent, but got a new one
```
RuntimeError: 401 Client Error: Unauthorized for url: https://python.zulipchat.com/api/v1/messages?anchor=oldest&num_before=0&num_after=2&apply_markdown=false&narrow=%5B%7B%22operator%22%3A+%22stream%22%2C+%22operand%22%3A+%22importlib%22%7D%5D
```

I was using the working token and email. I regenerated and tested with this [script](https://gist.github.com/vchrombie/bfd6d967e03bbe28cf4241995fb4b91b#file-script-py).

Did you have the same error?

vchrombie replied:
I understand that the argument `auth` should be passed to these functions, [client.py#L143 ](https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/client.py#L143 ) & [client.py#L174](https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/client.py#L174). Let me know if I am wrong.

But I see a `None` object passing through it. :thinking:  :confused:

valeriocos replied:
The method fetch at https://github.com/chaoss/grimoirelab-perceval/pull/667/files#diff-55fae40585a834540929c7f91a9c1e89R194, it isn't needed. The call at https://github.com/chaoss/grimoirelab-perceval/pull/667/files#diff-55fae40585a834540929c7f91a9c1e89R220 will use the parent `fetch`

vchrombie replied:
Sorry for the confusion, I got it right now. :+1:

valeriocos replied:
No worries, sorry for the mistake with the diffs in the first iteration

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,131 @@
+                                   help="Zulip chat URL")
+        parser.parser.add_argument('stream',
+                                   help='Zulip chat stream name')
+        parser.parser.add_argument('-e', '--email', dest='email',
```
Groupsio -> Zulip

vchrombie commented:
Thanks for the suggestions and pointers @valeriocos.

> I understand that the support for pagination is missing, please give it a try and let me know if you find any blocker. 

I have a logic in my mind. Once I fix this issue [#667 (review comment)](https://github.com/chaoss/grimoirelab-perceval/pull/667#discussion_r429567496), I will head to implement it and let you know about it. :smiley:

valeriocos commented:
Hi @vchrombie , I guess yesterday I make a mistake when copying the diffs, since the backend was working for me. From your current code you should apply the following diff:
```
diff --git a/perceval/backends/core/zulip.py b/perceval/backends/core/zulip.py
index 4d61a5a..0bf5b09 100644
--- a/perceval/backends/core/zulip.py
+++ b/perceval/backends/core/zulip.py
@@ -160,6 +160,15 @@ class Zulip(Backend):
         ts = float(item['timestamp'])
         return ts
 
+    @staticmethod
+    def metadata_category(item):
+        """Extracts the category from a Zulip item.
+
+        This backend only generates one type of item which is
+        'message'.
+        """
+        return CATEGORY_MESSAGE
+
     def _init_client(self, from_archive=False):
         """Init client"""
 
@@ -191,19 +200,6 @@ class ZulipClient(HttpClient):
 
         super().__init__(url, archive=archive, from_archive=from_archive, ssl_verify=ssl_verify)
 
-    def fetch(self, url, payload=None, headers=None, auth=None):
-        """Fetch the data from a given URL.
-
-        :param url: link to the resource
-        :param payload: payload of the request
-        :param headers: headers of the request
-        :param auth: auth of the request
-
-        :return: a response object
-        """
-        response = super().fetch(url, payload)
-        return response
-
     def get_messages(self, anchor):
         """Fetch the messages."""
```

vchrombie commented:
Thanks for the help @valeriocos. I got it right. :smiley:

Also, we have only `item_id` in the `search_fields`
```
    "search_fields": {
        "item_id": "159281179"
    },
```

I was thinking of adding `stream` too
```
    "search_fields": {
        "item_id": "159281179",
        "stream": "importlib"
    },
```
Also, do you think any more fields would go good ?

valeriocos commented:
You're welcome @vchrombie ! :)

> I was thinking of adding stream too

That's a good idea!

> Also, do you think any more fields would go good ?

The idea of the `search_fields` is to simplify the searches by avoiding to inspect what is inside the `data` attribute. For the moment the id and stream seem enough, maybe in the future we could add the `flags` or the `topic_links` if we see the contain valuable info.

vchrombie commented:
Hi @valeriocos 
I updated the PR according to your suggestions and it worked. I implemented the chain fetching too.

I have tested the backend and here are the results.

- python.zulipchat.com, stream: `importlib`
```
[2020-05-24 16:53:48,804] - Summary of results

	   Total items: 	12
	Items produced: 	12
	 Items skipped: 	0

	Last item UUID: 	a836795bdaf7a49c49f05424e4875651d059432c
	Last item date: 	2019-09-10 00:53:55+00:00

	Min. item date: 	2019-02-24 15:42:39+00:00
	Max. item date: 	2019-09-10 00:53:55+00:00

	Min. offset: 	-	Max. offset: 	-	Last offset: 	-


[2020-05-24 16:53:48,805] - Sir Perceval completed his quest.
```

- python.zulipchat.com, stream: `core/help`
```
[2020-05-24 15:53:49,517] - Summary of results

	   Total items: 	820
	Items produced: 	820
	 Items skipped: 	0

	Last item UUID: 	77fc711abe7ca9738d7c496b1efbcdf455b5e249
	Last item date: 	2020-04-27 01:38:46+00:00

	Min. item date: 	2018-04-05 00:49:12+00:00
	Max. item date: 	2020-04-27 01:38:46+00:00

	Min. offset: 	-	Max. offset: 	-	Last offset: 	-


[2020-05-24 15:53:49,518] - Sir Perceval completed his quest.
```

- anitab-org.zulipchat.com, stream: `design`
```
[2020-05-24 16:14:53,232] - Summary of results

	   Total items: 	1632
	Items produced: 	1632
	 Items skipped: 	0

	Last item UUID: 	c28b3fbdb777be873cc1934e73bbb603d0edaa58
	Last item date: 	2020-05-24 10:17:43+00:00

	Min. item date: 	2019-12-06 10:13:27+00:00
	Max. item date: 	2020-05-24 10:17:43+00:00

	Min. offset: 	-	Max. offset: 	-	Last offset: 	-


[2020-05-24 16:14:53,234] - Sir Perceval completed his quest.
```

There were no errors while fetching the chats.

I started working on adding the tests and other things need to get this PR completed.
Once it is done, I will squash the commits and fix the signoff. :grimacing: 

Thanks.

valeriocos commented:
Thank you for the update @vchrombie . Please ping me when the PR is ready for review.

vchrombie commented:
Hi @valeriocos 
I have worked on the tests but they are not completed. 

There are still some things pending. I have faced some errors. I will check on this by tomorrow.
Please leave any suggestions, if you have any when you are free.

valeriocos commented on tests/test_zulip.py:
```diff
@@ -0,0 +1,223 @@
+
+
+class TestZulipCommand(unittest.TestCase):
+    """Tests for GitterCommand class"""
```
typo GitterCommand

valeriocos commented on tests/test_zulip.py:
```diff
@@ -0,0 +1,223 @@
+    """Tests for GitterCommand class"""
+
+    def test_backend_class(self):
+        """Test if the backend class is Gitter"""
```
typo Gitter

valeriocos commented on tests/test_zulip.py:
```diff
@@ -0,0 +1,223 @@
+    def test_has_resuming(self):
+        """Test if it returns False when has_resuming is called"""
+
+        self.assertEqual(Zulip.has_resuming(), True)
```
you can use `assertTrue`

valeriocos commented on tests/test_zulip.py:
```diff
@@ -0,0 +1,223 @@
+    def test_has_archiving(self):
+        """Test if it returns True when has_archiving is called"""
+
+        self.assertEqual(Zulip.has_archiving(), True)
```
you can use `assertTrue`

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,257 @@
+        anchor = 'oldest'
+        # found_newest = False
+
+        while True:
```
Instead of using a while true and break, it would be better to control the loop with a variable to ease the readbility of the code

```
fetching = True

while fetching:

...

    if ...:
        fetching = False
```

vchrombie replied:
Okay.

vchrombie replied:
Can you check this once and let me know what do you think?

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,257 @@
+        # found_newest = False
+
+        while True:
+            raw_messages = self.client.get_messages(anchor)
```
Part of the code in the methods `get_anchor`, `found_newest` and `parse_messages` is duplicated, thus it may be convenient to merge them into a unique method, for instance:
```
def parse_messages(self, raw_messages):
        result = json.loads(raw_messages)

        found_newest = result.get('found_newest', False)
        messages = result.get('messages', None)
        anchor = messages[1]['id'] if messages else None

        return messages, found_newest, anchor
```

vchrombie replied:
Yes, this seems way better. :+1:

valeriocos commented on perceval/backends/core/zulip.py:
```diff
@@ -0,0 +1,257 @@
+
+        :return: a generator of items
+        """
+        anchor = 'oldest'
```
Does the anchor value change during the execution?

vchrombie replied:
Yes, initially the anchor is `oldest` and then we use the next message-id as the anchor to extract the next set of messages.

It will get two messages per each fetch (as it is [configured](https://github.com/vchrombie/grimoirelab-perceval/blob/zulip/perceval/backends/core/zulip.py#L221)). 

The API needs the `anchor` parameter and as we don't know the number of messages beforehand, I implemented this logic.

@valeriocos, do you have any comments?

vchrombie commented:
Thanks for the review @valeriocos.

> please remember to complete the docstrings and tests.

I planned to do it once I fix all the tests. I will fix them for sure. :+1: 

> If this can be of any help, you can run the coverage locally to make sure all code is covered

Thanks, I will use it.

vchrombie commented:
Adhering to the [contributing guidelines (#incubating-repositories)](https://github.com/chaoss/grimoirelab/blob/master/CONTRIBUTING.md#incubating-repositories), this work is moved to a separate repository [1] and will be maintained over there for some time. We can open this PR again and update it when we want to merge this backend into the main module.

[1] Zulip Backend for Perceval: https://github.com/vchrombie/grimoirelab-perceval-zulip

Best,
Venu

sduenas commented:
Thanks!
